plugins/fr.inria.diverse.melange.ui/src/main/java/fr/inria/diverse/melange/ui/scope/MelangeScopeProvider.xtend

```diff
@@ -0,0 +1,116 @@
+package fr.inria.diverse.melange.ui.scope
+
+import com.google.inject.Inject
+import fr.inria.diverse.melange.ast.ModelingElementExtensions
+import fr.inria.diverse.melange.metamodel.melange.ClassBinding
+import fr.inria.diverse.melange.metamodel.melange.Import
+import fr.inria.diverse.melange.metamodel.melange.Language
+import fr.inria.diverse.melange.metamodel.melange.MelangePackage
+import fr.inria.diverse.melange.metamodel.melange.Merge
+import fr.inria.diverse.melange.metamodel.melange.PackageBinding
+import fr.inria.diverse.melange.metamodel.melange.Slice
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.emf.ecore.EReference
+import org.eclipse.xtext.scoping.IScope
+import org.eclipse.xtext.scoping.Scopes
+import org.eclipse.xtext.scoping.impl.AbstractDeclarativeScopeProvider
+import org.eclipse.xtext.scoping.impl.SimpleScope
+import org.eclipse.xtext.ui.resource.XtextResourceSetProvider
+import org.eclipse.xtext.naming.IQualifiedNameProvider
+
+class MelangeScopeProvider extends AbstractDeclarativeScopeProvider{
+
+    @Inject extension ModelingElementExtensions
+    @Inject XtextResourceSetProvider rsProvider
+    @Inject IQualifiedNameProvider prov
+
+    def dispatch IScope getScope(EObject context, EReference reference) {
+        return super.getScope(context, reference)
+    }
+
+    def dispatch IScope getScope(Language language, EReference ref){
+        if (ref.featureID == MelangePackage.LANGUAGE__OPERATORS) {
+        	val owning = language.name
+            val basic = super.getScope((language as EObject), ref).allElements
+            val filtered = basic.filter[!it.name.toString.contains(owning)]
+            val scope = new SimpleScope(IScope.NULLSCOPE, filtered)
+
+            return scope
+        }
+        return super.getScope(language as EObject, ref)
+    }
+
+    def dispatch IScope getScope(Slice slice, EReference ref){
+        if (ref.featureID == MelangePackage.SLICE__ROOTS) {
+            val candidates = slice.slicedLanguage.syntax.pkgs
+            					.map[EClassifiers]
+            					.flatten
+            val  scope = Scopes.scopeFor(candidates)
+
+            return scope
+        }
+        else if (ref.featureID == MelangePackage.SLICE__MAPPING_RULES){
+        	val candidates = slice.slicedLanguage.syntax.pkgs
+        	val scope = Scopes.scopeFor(candidates)
+
+        	return scope
+        }
+        return super.getScope(slice as EObject, ref)
+    }
+
+    def dispatch IScope getScope(Merge merge, EReference ref){
+    	if(ref.featureID == MelangePackage.MERGE__MERGED_LANGUAGE) {
+    		val owning = merge.owningLanguage
+    		val basic = super.getScope((owning) as EObject, ref).allElements
+    		val filtered = basic.filter[!it.name.toString.contains(owning.name)]
+    		val scope = new SimpleScope(IScope.NULLSCOPE, filtered)
```

Comment:
Maybe use the factory instead? Scopes::scopeFor(Iterable<? extends EObject> elements)